### ISSUE TITLE
[Feature Request] option to squash for propose

### ISSUE THREAD

**Issue by bryanlarsen** [git-town/git-town#3077]
I created a nice stack of 3 feature branches in git town and had an awesome experience during the dev phase.   It was really nice keeping the stack rebased on top of incoming changes on main during dev, and I really liked how it looked in GitLab when I `propose`'d all 3.    And it was really nice to switch and sync between branches while I was tweaking things so they all had a green CI run.

Unfortunately, the experience from there wasn't so clean.   Reviewers approved the top 2 PR's and had a comment on the 3rd.   We used the squash merge button in GitLab to merge the first PR.   We then encountered the first problem.   Because the squash changed history, we weren't able to immediately click on the "Merge" button in Gitlab to merge the 2nd one immediately.    We had to click on the "Rebase" button and the "Merge when pipeline succeeds" button.   And then another PR got merged between them and we had to redo it.

The second annoyance was when I `git town switch`'d back to the 3rd PR after the first two had been merged.   Because the first 2 branches had been merged as squashes, I got a whole bunch of rebase conflicts as it tried to apply the original individual patches of those 2 PR's.

Alternative request A:  somebody to tell me "there's a better way of doing it".

Alternative request B:  But if I am doing it right, it seems to me that if `propose` did the squash rather than letting GitLab do it, I wouldn't have run into either stumbling block.

Alternative request C:  I like having the individual unsquashed commits visible in GitLab.   Now if somebody looks at the git log, they can click on the linked issue number and then click on the PR from the issue and see the full history with it's "WIP's" and "WTF's" and "lint fix" etc.   The git log is clean, but the "real" history is accessible.   Best of both worlds!

It seems to me we can this best of both worlds with a 4 step process:   create the PR as a draft without squash, squash, and then force push and then remove the draft flag.

For my next stack I'm going to try doing that process manually.

**Comment by kevgo** [comment]
> We used the squash merge button in GitLab to merge the first PR. [...] Because the squash changed history, we weren't able to immediately click on the "Merge" button in Gitlab to merge the 2nd one immediately. We had to click on the "Rebase" button and the "Merge when pipeline succeeds" button.

The commits in your 2nd branch are based on the commits on your main branch and the commits in the first branch. When you squash-merge the first branch into main, the commits on main change, and the first branch gets thrown away. Hence you need to re-base the commits on the second branch before you can merge them. Git is just doing the right thing here.

Here are a few ways to avoid this problem:

Option A: Instead of squash-merging, compress all commits in each feature branch down to one commit (via interactive rebase or Git Town's planned [compress command](https://github.com/git-town/git-town/issues/1529)) and configure the PR to merge using [fast-forward merge](https://docs.gitlab.com/ee/user/project/merge_requests/methods/#fast-forward-merge) or [merge commit with semi-linear history](https://docs.gitlab.com/ee/user/project/merge_requests/methods/#merge-commit-with-semi-linear-history) in GitLab when creating the merge request. This moves the commit from the first branch onto the main branch, and you hopefully won't need to rebase your other branches.

Option B: Avoid stacking branches: implement your change as independent top-level feature branches instead. Instead of this:

```
main
 \
  first branch (child of main)
   \
    second branch (child of first)
     \
      third branch (child of second)
```

you do this:

```
main
 \
  first branch (child of main)
 \
  second branch (child of main)
 \ 
  third branch (child of main)
```

That's not always possible (hence stacking was invented), but it's often possible. Another benefit is that you can now merge branch 1, 2, and 3 in any order.

**Comment by kevgo** [comment]
> The second annoyance was when I git town switch'd back to the 3rd PR after the first two had been merged. Because the first 2 branches had been merged as squashes, I got a whole bunch of rebase conflicts as it tried to apply the original individual patches of those 2 PR's.

You can minimize this problem by running these steps:

1. Run `git sync --all` right before merging a branch to get your workspace fully synced with the server.
2. Right after that squash-merge the PR on GitLab
3. Right after that run `git sync --all` again.

You will encounter merge conflicts, and if nobody else shipped anything in the meantime, you can just accept the current changes in each branch by running `git checkout --ours`. I use VSCode, which has nice features to resolve merge conflicts, and click "Accept Current Change" on each merge conflict.

**Comment by kevgo** [comment]
> if I am doing it right, it seems to me that if propose did the squash rather than letting GitLab do it, I wouldn't have run into either stumbling block.

"git propose" only opens the URL for creating a new merge request. You then create the merge requests yourself, using the configuration you want.

**Comment by kevgo** [comment]
> Alternative request C: I like having the individual unsquashed commits visible in GitLab. Now if somebody looks at the git log, they can click on the linked issue number and then click on the PR from the issue and see the full history with it's "WIP's" and "WTF's" and "lint fix" etc. The git log is clean, but the "real" history is accessible. Best of both worlds!

I think that's what the [merge commit](https://docs.gitlab.com/ee/user/project/merge_requests/methods/#merge-commit) option for merge requests does. You can configure it in the web page that `git propose` opens in your browser.

**Comment by kevgo** [comment]
> It seems to me we can this best of both worlds with a 4 step process: create the PR as a draft without squash, squash, and then force push and then remove the draft flag.

I don't quite understand where you are getting at. My proposal is:

1. Create a normal merge request and develop normally (committing lots of "WIP's" and "WTF's" and "lint fix").
2. When your branch is ready for review, squash all commits on the branch (all the "WIP's" and "WTF's" and "lint fix") into a single commit.
3. Create a merge request using the [fast-forward merge](https://docs.gitlab.com/ee/user/project/merge_requests/methods/#fast-forward-merge) option. This will result in a single commit on the main branch, similar to the squash-merge option you used before.
4. Merge the merge request normally. You should run into a lot less of the issues you described above.

**Comment by bryanlarsen** [comment]
Missing pieces of information in original post:

- The CI process takes an hour.   Yes, there are places it can be optimized (and we're working on that), but there are some excellent integration tests in the suite that would be difficult to speed up significantly.
- Due to SOC2 we need to have a successful CI on the exact git hash before merging an MR
- We prefer to review the MR pre-squash

So let's take an example day.   There are 12 MR's to merge.   Some of the MR's are dependent on previous MR's, but most are independent and conflict free.  Taking 12 hours to merge 12 MR's is unacceptable.

So we need a procedure such that the final CI runs can be done in parallel instead of in series.   We employ the concept of what we call a "merge train".   Once an MR is reviewed and approved, it is "added to the merge train".   IOW, it is squashed and rebased onto the end of the merge train. 

So the merge train might look like this:

```
main
 \
  A, 10 minutes left on final CI
   \
    B, 20 minutes left on final CI
     \
      C, 30 minutes left on final CI.   Also tagged as "merge-train"
```

So when D gets reviewed, if somebody clicks the "rebase onto main" and "merge after CI completes" buttons, the merge will get cancelled when A merges into main.  Instead, if D is squashed and rebased onto C, by the time that D's final CI is complete C will be main and D can be merged immediately.

**Comment by kevgo** [comment]
Have you tried the [fast-forward merges](https://docs.gitlab.com/ee/user/project/merge_requests/methods/#fast-forward-merge) together with feature branches that consist of only one commit? The documentation says:

> This method is equivalent to `git merge --ff <source-branch>` for regular merges.

This sounds to me like exactly what you need. If each feature branch consists of only one commit, then the merge would mean to simply point the main branch to the commit in the feature branch. No SHAs change. No need to re-test main after the merge.

With the right setup, the merge train could merge branch A from your example after 10 minutes, branch B after 20 minutes, and branch C after 30 minutes (assuming they are all green).

I don't know the GitLab merge train feature too well but I would be surprised if it doesn't support this. It's a very common use case and many other GitLab users have the exact same problem that you describe.

**Comment by bryanlarsen** [comment]
Yes, that's what I'm setting up for GitLab to take advantage of.   It's a tedious manual error prone process.   If we were on GitHub we could buy Graphite, but we're not so I was hoping to use git town instead.   Issue #2223 makes it particularly painful and the main reason why I've stopped using Git Town for this.   GitLab has no concept of a "merge train", it's all manual work.

After a merge request is approved I have to:

1. squash it
2. rebase it onto the end of the merge train
3. force push
4. edit the merge request to change the parent
5. edit the merge request to remove the squash flag  (because if I don't GitLab will change the commit message on merge, changing the hash).
6. wait for CI to finish
7. click the merge button

(Instead of 5&6 I can wait for the previous MR's CI to finish and then click the "merge when pipeline succeeds".   I can't click that button immediately because if a pipeline up the chain fails but a mine succeeds we get the two MR's merged together)

It feels like git town is so close.    It already knows how to do step 2 & step 4, which seems like the harder parts.   Neither 4 nor 5 can be done with the glab CLI.

**Comment by bryanlarsen** [comment]
It looks like your steps 1 & 3 are incompatible on GitLab: https://gitlab.com/groups/gitlab-org/-/epics/4911.    Hopefully it lands soon.

The other problem is that I'm hitting #2223 on using `git rebase -i` to do a squash.   `git rebase -i` can move commits around so you have to ensure that you force push all affected branches, not just the top. I think I've figured it out now, but it sure frustrated me until I did.

**Comment by kevgo** [comment]
> It looks like your steps 1 & 3 are incompatible on GitLab: https://gitlab.com/groups/gitlab-org/-/epics/4911. Hopefully it lands soon.

Wow. I'm surprised that GitLab's merge train doesn't support fast forward merges since merge trains make a lot less sense without fast-forward merges. The person-years spent discussing this feature (the issue is 4 years old, has 100 participants, and when printed in a small font its 54 pages) would have been enough to add fast-forward merges at least 10 times. The feature already exists and works for normal merges. Not sure what they are waiting for here.

> git rebase -i can move commits around so you have to ensure that you force push all affected branches, not just the top. I think I've figured it out now, but it sure frustrated me until I did.

To help evaluate if there is something Git Town could do to help with syncing after a `rebase -i`, can you please outline what you did to resolve the problem? Just force-push the entire branch stack? You have given me inspiration for #3177. Curious what you think about this idea!

**Comment by kevgo** [comment]
Git Town v14 provides the [compress](https://www.git-town.com/commands/compress) command. It allows you to squash commits of a feature branch or even a stack of feature branches. Given that this addresses the original requests of this ticket, I'm going to close this ticket for the time being. If there is additional feedback, please comment and/or re-open it. Thanks!
